sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -364,48 +366,162 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
+    val tableIdentWithDB = tableIdent.copy(database = Some(db))
+    val tableName = tableIdentWithDB.unquotedString
+
+    catalog.getTableMetadataOption(tableIdentWithDB) match {
+      // If the table already exists...
+      case Some(existingTable) =>
+        mode match {
+          case SaveMode.Ignore => // Do nothing
+
+          case SaveMode.ErrorIfExists =>
+            throw new AnalysisException(s"Table $tableName already exists. You can set SaveMode " +
+              "to SaveMode.Append to insert data into the table or set SaveMode to " +
+              "SaveMode.Overwrite to overwrite the existing data.")
+
+          case SaveMode.Append =>
+            if (existingTable.tableType == CatalogTableType.VIEW) {
+              throw new AnalysisException("Saving data into a view is not allowed.")
+            }
+
+            if (existingTable.provider.get == DDLUtils.HIVE_PROVIDER) {
+              throw new AnalysisException(s"Saving data in the Hive serde table $tableName is " +
+                "not supported yet. Please use the insertInto() API as an alternative.")
+            }
+
+            // Check if the specified data source match the data source of the existing table.
+            val existingProvider = DataSource.lookupDataSource(existingTable.provider.get)
+            val specifiedProvider = DataSource.lookupDataSource(source)
+            // TODO: Check that options from the resolved relation match the relation that we are
+            // inserting into (i.e. using the same compression).
+            if (existingProvider != specifiedProvider) {
+              throw new AnalysisException(s"The format of the existing table $tableName is " +
+                s"`${existingProvider.getSimpleName}`. It doesn't match the specified format " +
+                s"`${specifiedProvider.getSimpleName}`.")
+            }
+
+            if (df.schema.length != existingTable.schema.length) {
+              throw new AnalysisException(
+                s"The column number of the existing table $tableName" +
+                  s"(${existingTable.schema.catalogString}) doesn't match the data schema" +
+                  s"(${df.schema.catalogString})")
+            }
+
+            val resolver = df.sparkSession.sessionState.conf.resolver
+            val tableCols = existingTable.schema.map(_.name)
+
+            // As we are inserting into an existing table, we should respect the existing schema and
+            // adjust the column order of the given dataframe according to it, or throw exception
+            // if the column names do not match.
+            val adjustedColumns = tableCols.map { col =>
+              df.queryExecution.analyzed.resolve(Seq(col), resolver).getOrElse {
+                val inputColumns = df.schema.map(_.name).mkString(", ")
+                throw new AnalysisException(
+                  s"cannot resolve '$col' given input columns: [$inputColumns]")
+              }
+            }
```

Comment:
Do we still need the comment at https://github.com/apache/spark/pull/15996/files#diff-945e51801b84b92da242fcb42f83f5f5L195 ?